Refactor Nalgebra integration

jamwaffles opened the pull request:
* Drastically improves docs for `Coord` and `UnsignedCoord` (even if they're changed a bunch later)
* Changes Nalgebra integration from a type alias (`type Coord = nalgebra::Vector2<i32>` to implementing `From<Vector2<N>>` for `Coord` and `UnsignedCoord`. This allows for less breakage between non-`nalgebra_support` and `nalgebra_support` modes but introduces a breaking change when converting between Nalgebra-using code and e-g
* Closes #153

rfuest commented on embedded-graphics/src/coord.rs:
```diff
@@ -2,184 +2,230 @@
+/// 2D signed integer coordinate type
+///
+/// This coordinate should be used to define graphics object coordinates. For example, a
+/// [`Rect`] may be defined that has its top left at `(-1,-2)`. To specify positive-only screen
```
Nit: `Rect` -> `Rectangle`

rfuest commented on embedded-graphics/src/coord.rs:
````diff
@@ -2,184 +2,230 @@
+/// ```
+///
+/// [`UnsignedCoord`]: ../unsignedcoord/struct.UnsignedCoord.html
+/// [`Rect`]: ../primitives/rectangle/struct.Rectangle.html
````
Nit: `Rect` -> `Rectangle`

rfuest commented on embedded-graphics/src/coord.rs:
```diff
@@ -2,184 +2,230 @@
+
+impl Coord {
+    /// Create a new coordinate with X and Y values
+    pub fn new(x: i32, y: i32) -> Self {
```
This could be a `const fn`.

rfuest commented on embedded-graphics/src/coord.rs:
```diff
@@ -2,184 +2,230 @@
-    /// `Coord` instead of this builtin implementation.
-    ///
-    /// [`UnsignedCoord`]: ../unsignedcoord/struct.UnsignedCoord.html
-    /// [`Rect`]: ../primitives/rectangle/struct.Rectangle.html
```
Nit: `Rect` -> `Rectangle`

rfuest commented on embedded-graphics/src/coord.rs:
```diff
@@ -2,184 +2,230 @@
+        match idx {
+            0 => &self.0,
+            1 => &self.1,
+            _ => panic!("Unreachable index {}", idx),
```
Nit: Doesn't "Unreachable index" sound a little stange?
The rust compiler uses the following error for `array[n]`: "index out of bounds: the len is 2 but the index is 2".

rfuest commented on embedded-graphics/src/coord.rs:
```diff
@@ -2,184 +2,230 @@
-        }
+impl From<(u32, u32)> for Coord {
+    fn from(other: (u32, u32)) -> Self {
+        Self(other.0 as i32, other.1 as i32)
```
What happens if someone calls this with `u32::max_value()`?
There are a lot of `From<...>` implementations, are all of these really necessary?

jamwaffles replied:
Excellent point. None of these `u32` conversions should've been implemented for `Coord` in the first place. Might've just been some bad copypasta. At any rate, I've removed all `u32`-based conversions for `Coord`.

rfuest commented on embedded-graphics/src/unsignedcoord.rs:
```diff
@@ -2,124 +2,156 @@
+
+impl UnsignedCoord {
+    /// Create a new coordinate with X and Y values
+    pub fn new(x: u32, y: u32) -> Self {
```
Could be a `const fn`.

rfuest commented on embedded-graphics/src/unsignedcoord.rs:
```diff
@@ -2,124 +2,156 @@
+        match idx {
+            0 => &self.0,
+            1 => &self.1,
+            _ => panic!("Unreachable index {}", idx),
```
See `Index` impl for `Coord`.

jamwaffles commented:
Thanks for the review! I've addressed all comments with some changes - what you've raised all makes sense to me.

As for all the `From<>` impls, I think it's useful to provide that compatibility. I realise they expand the API surface for this crate and might become a maintenance burden, but for now I think it's good to keep the compatibility the impls bring.
